### PR TITLE
Refactor: Single LocationOfInterestModel

### PR DESCRIPTION
**Pull request by scolsen**
This commit simplifies and updates our model for locations of interest:

- We retain only a single LocationOfInterest data class that accepts a
  geometry in its constructor.
- The Builder for this class has been removed.
- We now return the type of the LocationOfInterest based on its
  geometry.
- The AreaOfInterest and PointOfInterest classes have been removed.

In addition, I've added some extensions for working with the Point model
while it still exists, and I've converted FakeData to kotlin.

### CODE REVIEW

File: ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt:52]: Do we need to have a separate LOI type? Can we not just differentiate using `geometry.geometryType`?

File: ground/src/main/java/com/google/android/ground/model/locationofinterest/Point.java
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/model/locationofinterest/Point.java:26]: Should we instead be using JTS `Point` instead? Or should we just create our own topology classes in the model as I'm doing in https://github.com/google/ground-platform/pull/928? Sorry to flip-flop, would like to think about it well now since we won't have time to revisit for awhile.

Note that going forward we may want to consider keeping conversion logic separate from data classes.

File: ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt:49]: Looking at this now, wondering if it would be more type-safe to keep the AOI and POI classes, having AOI contain `Polygon` and POI contain `Point`?

**Review by gino-m — CHANGES_REQUESTED** [review]
A few questions/discussions.

File: ground/src/main/java/com/google/android/ground/model/locationofinterest/Point.java
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/model/locationofinterest/Point.java:26]: > Should we instead be using JTS Point instead? 

Yes. I decided to omit that from this PR, but the next one will delete this pointless `Point` class and use JTS directly.

But, if the broader plan is to use our own model, I think we should just do that. There's some benefit to having the representation on both sides be equivalent.

File: ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt:49]: It is more type safe, but we'd also be unable to construct LOIs, which is fine if we're ok with keeping that class abstract.

File: ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt:52]: Yes we could!

File: ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt:83]: Would this create a new instance of `LocationOfInterest` every time the `geometry` is updated? Is it not possible to update the property from the caller place?

File: ground/src/main/java/com/google/android/ground/persistence/remote/DataStoreException.kt
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/persistence/remote/DataStoreException.kt:42]: Can we use string template here as well?

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiConverter.kt
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiConverter.kt:64]: Instead of repeating the constructor code, can we use `.map()` on line 47 to convert to the appropriate geometry?

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionCollectionReference.kt
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionCollectionReference.kt:39]: Isn't this incorrect? Loi ID should never be empty.

File: ground/src/main/java/com/google/android/ground/repository/SubmissionRepository.kt
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/repository/SubmissionRepository.kt:160]: Should loi.id be non-nullable?

File: ground/src/main/java/com/google/android/ground/repository/SubmissionRepository.kt
Comment (shobhitagarwal1612) [inline ground/src/main/java/com/google/android/ground/repository/SubmissionRepository.kt:183]: Same as above.

**Review by shobhitagarwal1612 — CHANGES_REQUESTED** [review]

File: ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt:83]: Yes that's right, it always returns a new instance. This is generally how you update immutable structures in functional programming. The big difference here is that functional languages usually use sharing under the hood to mitigate cost (so even though it looks like an immutable copy, it's implemented as mutation under the hood).

I'm not sure whether or not Kotlin has similar optimizations, we'd have to look at the bytecode. Instead, I can change this to a `var` and make the setter private to give us immutability without the additional allocation costs.

**Review by gino-m — APPROVED** [review]
LGTM, please resolve Shobhit's comments where relevant before merging.

File: ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt
Comment (gino-m) [reply ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt:49]: We spoke. A single concrete LOI class with `Geometry` will work; we can specialize into subclasses iff we need to down the road.

File: ground/src/main/java/com/google/android/ground/model/locationofinterest/Point.java
Comment (gino-m) [reply ground/src/main/java/com/google/android/ground/model/locationofinterest/Point.java:26]: We spoke. Let's go with our own geometry impl, but please feel free to merge this PR first and retrofit.

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiConverter.kt
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiConverter.kt:52]: One more reason to create our own model classes (runtime String bindings are more error prone as opposed to static enums)

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiConverter.kt
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiConverter.kt:59]: Please rename `fillLocationOfInterest` to `createLocationOfInterest` or similar to make its new scope clear.

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionCollectionReference.kt
Comment (gino-m) [reply ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionCollectionReference.kt:39]: +1. Though there are many cases where we could benefit from making model attributes non-null; we have a TODO in our planning doc to fix so feel free to address in a separate PR.

File: ground/src/main/java/com/google/android/ground/repository/SubmissionRepository.kt
Comment (gino-m) [inline ground/src/main/java/com/google/android/ground/repository/SubmissionRepository.kt:148]: Nit: Imo short-circuiting is still more readable here (eg. `if (==null) return error; return applyAndEnqueue...`), mostly since the `applyAndEnqueue` expression is non-trivial. Is there any official style guidance on this?

**Review by gino-m — CHANGES_REQUESTED** [review]
Undo, LGTM. Added a few more minor comments.

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiConverter.kt
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiConverter.kt:64]: great idea!

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiConverter.kt
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiConverter.kt:59]: done!

File: ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/model/locationofinterest/LocationOfInterest.kt:52]: I'll hold off on this for now since we're going to roll our own Geometry model

File: ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionCollectionReference.kt
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/SubmissionCollectionReference.kt:39]: sg, let's tackle this separately.

File: ground/src/main/java/com/google/android/ground/repository/SubmissionRepository.kt
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/repository/SubmissionRepository.kt:183]: Thanks! Will address in a follow up.

Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/repository/SubmissionRepository.kt:183]: oh- scratch that, this field is already non-null in this PR, these checks are redundant. Removed.

File: ground/src/main/java/com/google/android/ground/repository/SubmissionRepository.kt
Comment (scolsen) [reply ground/src/main/java/com/google/android/ground/repository/SubmissionRepository.kt:148]: the check was actually just redundant, so it's gone.

**Comment by scolsen** [comment]
@gino-m ready for another look!

**Review by gino-m — APPROVED** [review]
LGTM, great work! Looking forward to our own model reducing some of the complexity further!

**Review by shobhitagarwal1612 — APPROVED** [review]

**Pull request merged**
